File: app/src/main/java/com/shen/xi/android/tut/QingImageViewActivity.java
package com.shen.xi.android.tut;

import android.os.Bundle;
import android.view.MenuItem;

import com.google.api.client.json.JsonParser;
import com.google.api.client.repackaged.com.google.common.base.Strings;
import com.shen.xi.android.tut.sinablog.ArticleInfo;

import java.io.IOException;
import java.util.List;

import static com.shen.xi.android.tut.MainActivity.ImageSource.QingTag;
import static com.shen.xi.android.tut.MainActivity.ImageSource.Unknown;


public class QingImageViewActivity extends AbstractImageViewActivity {

    public static final String QING_SOURCE = "qing source";
    private List<ArticleInfo> mArticleInfoList;
    private MainActivity.ImageSource mSource = Unknown;
    private List<String> mImageUrlList;

    public QingImageViewActivity() {
        super(R.menu.qing_image_view);
        TuTModule.getInjector().injectMembers(this);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle extras = getIntent().getExtras();
        int item = extras.getInt(ITEM_POSITION, 0);
        String sourceString = extras.getString(QING_SOURCE);
        String jsonList = extras.getString(JSON_LIST);

        if (!Strings.isNullOrEmpty(sourceString)) {
            try {
                mSource = MainActivity.ImageSource.valueOf(sourceString);
            } catch (IllegalArgumentException e) {
                mSource = Unknown;
            }
        }

        ImagePagerAdapter imagePagerAdapter;
        int size = 0;
        try {
            JsonParser jsonParser = mJsonFactory.createJsonParser(jsonList);
            switch (mSource) {
                case QingTag:
                    mArticleInfoList = (List<ArticleInfo>)
                            jsonParser.parseArray(List.class, ArticleInfo.class);
                    size = mArticleInfoList.size();
                    break;

                case QingPage:
                    mImageUrlList = (List<String>) jsonParser.parseArray(List.class, String.class);
                    size = mImageUrlList.size();
                    break;

            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (size != 0) {
            imagePagerAdapter = new ImagePagerAdapter(getSupportFragmentManager(), size);
            setupPager(imagePagerAdapter, item);
        } else {
            finish();
        }
    }

    @Override
    protected String getImageUrlByOrder(int order) {
        return mSource == QingTag
                // switch to the high-def version
                ? mArticleInfoList.get(order).imageSrc.replace("mw205", "mw600")
                : mImageUrlList.get(order);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        return id == R.id.action_settings || super.onOptionsItemSelected(item);
    }


}

File: app/src/main/scala/com/shen/xi/android/tut/QingImageViewActivity.scala
package com.shen.xi.android.tut


import java.util.{List => JList}

import android.os.Bundle
import android.view.MenuItem
import com.google.common.base.Strings
import com.shen.xi.android.tut.sinablog.ArticleInfo


object QingImageViewActivity {
  val QING_SOURCE = "qing source"
  val QING_TITLE = "qing title"
}

class QingImageViewActivity extends AbstractImageViewActivity(R.menu.qing_image_view) {

  import com.shen.xi.android.tut.AbstractImageViewActivity._
  import com.shen.xi.android.tut.QingImageViewActivity._

  private var mArticleInfoList: JList[ArticleInfo] = null
  private var mSource: ImageSource = Unknown
  private var mImageUrlList: JList[String] = null

  TuTModule.getInjector.injectMembers(this)

  override def onCreate(savedInstanceState: Bundle) = {
    super.onCreate(savedInstanceState)

    val extras = getIntent.getExtras
    val item = extras.getInt(ITEM_POSITION, 0)
    val sourceString = extras.getString(QING_SOURCE)
    val jsonList = extras.getString(JSON_LIST)
    val jsonParser = mJsonFactory.createJsonParser(jsonList)
    var size = 0
    setTitle(extras.getString(QING_TITLE))

    if (!Strings.isNullOrEmpty(sourceString)) {
      try {
        mSource = ImageSource.fromName(sourceString)
      } catch {
        case e: IllegalArgumentException => mSource = Unknown
      }
    }



    mSource match {
      case QingTag =>
        mArticleInfoList = jsonParser.parseArray[ArticleInfo](classOf[JList[ArticleInfo]], classOf[ArticleInfo]).asInstanceOf[JList[ArticleInfo]]
        size = mArticleInfoList.size

      case QingPage =>
        mImageUrlList = jsonParser.parseArray[String](classOf[JList[String]], classOf[String]).asInstanceOf[JList[String]]
        size = mImageUrlList.size
    }

    if (size != 0) {
      val imagePagerAdapter = new ImagePagerAdapter(getSupportFragmentManager, size)
      setupPager(imagePagerAdapter, item)
    } else {
      finish()
    }
  }

  override def getImageUrlByOrder(order: Int) = mSource match {
    // switch to the high-def version
    case QingTag => mArticleInfoList.get(order).href.replace("mw205", "mw600")
    case QingPage => mImageUrlList.get(order)
  }

  // Handle action bar item clicks here. The action bar will
  // automatically handle clicks on the Home/Up button, so long
  // as you specify a parent activity in AndroidManifest.xml.
  override def onOptionsItemSelected(item: MenuItem) = item.getItemId match {
    case R.id.action_settings => true
    case _ => super.onOptionsItemSelected(item)
  }

}
